### PR TITLE
Python script for generating vector tests

### LINKED ISSUE #186: Vector Test Generator
Write a python script to automatically generate vector json tests. Add support for the script to be able to parse Mavis as well to generate the tests.

### PR DESCRIPTION
**Pull request by ADI-ROXX**
Resolves #186 

This PR provides a python script that generates vector tests automatically from the Mavis definitions.

### CODE REVIEW

**Review by klingaard — COMMENTED** [review]
This is very cool.  Curious, though, what's the pass/fail criteria?  It is as simple as olympia ran the test?

I'd like to propose a few changes:
1. Do we need the generated tests committed?  How about they are generated during the configuration of olympia?  I.e. they are created during the `cmake` invocation.
2. Can these tests be added the `regress` target for the simulator?  Specifically, add a `glob` test

**Comment by ADI-ROXX** [comment]
> What's the pass/fail criteria? It is as simple as olympia ran the test?

Yep. It is as simple as 'Olympia ran the tests' without any crashes. 
The current tests provide smoke testing across almost all vector logic cases.

> Do we need the generated tests committed?

Nope, not at all since they're generated by the python file. The generated tests are for reference of reviewer only. I will be removing the whole folder in the final commit after review completion.

> Can these tests be added the regress target for the simulator? Specifically, add a glob test

Sure, I'll add that right away.

**Comment by ADI-ROXX** [comment]
There is one issue that I am seeing: I don't know if it is global or just on my system:
```
adi@adi:~/Desktop/riscv-perf-model/build/test$ cd /home/adi/Desktop/riscv-perf-model/build/test && ctest -R Vector_test_gen_add -V
UpdateCTestConfiguration  from :/home/adi/Desktop/riscv-perf-model/build/test/DartConfiguration.tcl
Parse Config file:/home/adi/Desktop/riscv-perf-model/build/test/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/adi/Desktop/riscv-perf-model/build/test/DartConfiguration.tcl
Parse Config file:/home/adi/Desktop/riscv-perf-model/build/test/DartConfiguration.tcl
Test project /home/adi/Desktop/riscv-perf-model/build/test
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 128
    Start 128: Vector_test_gen_add

128: Test command: /home/adi/Desktop/riscv-perf-model/build/test/core/vector/Vector_test "-l" "top" "info" "gen_add.out" "-c" "test_cores/test_big_core.yaml" "--input-file" "add.json"
128: Working Directory: /home/adi/Desktop/riscv-perf-model/build/test/core/vector
128: Test timeout computed to be: 1500
128:   [in] Configuration: Node "" <- file: "test_cores/test_big_core.yaml"
128: 
128: Setting up Simulation Content...
128:   [out] placing tap on node top for: Tap location_pattern="top" (category="info") -> file: "gen_add.out"
128: Resources:
128:   cpu
128: Building tree...
128: Configuring tree...
128: Finalizing tree...
128: Inst Allocator: 1238 Inst objects allocated/created
128: Inst Allocator: 1238 Inst objects allocated/created
128: Inst Allocator: 0 Inst objects allocated/created
128: Inst Allocator: 1238 Inst objects allocated/created
128: Inst Allocator: 0 Inst objects allocated/created
128: Preparing to run...
128: Meta-Parameters:
128:   architecture: NONE
128:   is_final_config: false
128: Non-default model parameters: 24
128: Running...
128: olympia: JSON file input detected
128: Exception while running
128: total_valid_ != capacity(): FetchQueue: Queue is full: in file: '/usr/local/include/sparta/resources/Queue.hpp', on line: 703
128:   [out] Writing error dump file 'error-dump.dbg'
128:   [out] Debug state written to "error-dump.dbg"
128: terminate called after throwing an instance of 'sparta::SpartaException'
128:   what():  total_valid_ != capacity(): FetchQueue: Queue is full: in file: '/usr/local/include/sparta/resources/Queue.hpp', on line: 703
128: Received Signal 6 (Aborted), address is 0x3e800010696 from (nil)
128: ( 1) /home/adi/Desktop/riscv-perf-model/build/test/core/vector/Vector_test(+0x5cb9a0) [0xc00ed81eb9a0]
128: ( 2) linux-vdso.so.1(__kernel_rt_sigreturn+0) [0xeedac9a088f8]
128: ( 3) /lib/aarch64-linux-gnu/libc.so.6(+0x87608) [0xeedac90a7608]
128: ( 4) /lib/aarch64-linux-gnu/libc.so.6(gsignal+0x1c) [0xeedac905cb3c]
128: ( 5) /lib/aarch64-linux-gnu/libc.so.6(abort+0xf4) [0xeedac9047e00]
128: ( 6) /lib/aarch64-linux-gnu/libstdc++.so.6(__gnu_cxx::__verbose_terminate_handler()+0x1d8) [0xeedac937d4d8]
128: ( 7) /lib/aarch64-linux-gnu/libstdc++.so.6(+0xaa570) [0xeedac937a570]
128: ( 8) /lib/aarch64-linux-gnu/libstdc++.so.6(std::unexpected()+0) [0xeedac9370e04]
128: ( 9) /lib/aarch64-linux-gnu/libstdc++.so.6(__cxa_rethrow+0x64) [0xeedac937a984]
128: (10) /home/adi/Desktop/riscv-perf-model/build/test/core/vector/Vector_test(+0x4339cc) [0xc00ed80539cc]
128: (11) /home/adi/Desktop/riscv-perf-model/build/test/core/vector/Vector_test(+0x10a558) [0xc00ed7d2a558]
128: (12) /home/adi/Desktop/riscv-perf-model/build/test/core/vector/Vector_test(+0xae680) [0xc00ed7cce680]
128: (13) /lib/aarch64-linux-gnu/libc.so.6(+0x284c4) [0xeedac90484c4]
128: (14) /lib/aarch64-linux-gnu/libc.so.6(__libc_start_main+0x98) [0xeedac9048598]
128: (15) /home/adi/Desktop/riscv-perf-model/build/test/core/vector/Vector_test(+0xffe30) [0xc00ed7d1fe30]
128: Exiting
128: WARNING: Seems that not all of the blocks made it back.  
128: 'sparta::SpartaSharedPointerAllocator<PointerT>::~SpartaSharedPointerAllocator() [with PointerT = olympia::InstGroup]'
128: Allocated: 9
128: Returned: 2
1/1 Test #128: Vector_test_gen_add ..............***Failed    0.08 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.09 sec

The following tests FAILED:
        128 - Vector_test_gen_add (Failed)
Errors while running CTest
Output from these tests are in: /home/adi/Desktop/riscv-perf-model/build/test/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```


Summary: Fetch Queue overflow. Any json that has more than 10 tests will cause a Queue overflow.

**Comment by ADI-ROXX** [comment]
To solve this, I am thinking of batching the test json files in sub-folders. What do you think about it?

@klingaard Will this be okay?

**Comment by ADI-ROXX** [comment]
@klingaard One question though ( unrelated to this PR) for [LFX](https://mentorship.lfx.linuxfoundation.org/project/1dbecd05-87ae-476f-9d0a-1985cc001ec6), in the cover letter, do we have to also add a proposal on how are we proposing to solve the problem? Like, implementing a simplified demo solution ( with stubbed code ) showing how we plan to implement that.

File: scripts/generate_vector_tests.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    def generate_tests(self, output_dir):
+        output_path = Path(output_dir)
+        output_path.mkdir(parents=True, exist_ok=True)
+        
+        groups = defaultdict(list)
+        for inst in self.vector_instructions:
+            mnemonic = inst.get("mnemonic", "")
+            inst_type = inst.get("type", [])
+            
+            if isinstance(inst_type, list):
```
Comment (kathlenemagnus) [inline scripts/generate_vector_tests.py:169]: @ADI-ROXX how did you confirm that your groups were exclusive? Or did you come across any instructions that could be in multiple groups?

For loads and stores, it'd be great to break these groups up into multiple groups for the different addressing modes (unit stride, strided and indexed) and types (normal, segmented, whole register).

File: scripts/generate_vector_tests.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                elif "permute" in mnemonic or ("mv" in mnemonic and not mnemonic.startswith("vset")):
+                    group = "permute"
+                else:
+                    group = "other"
+            else:
+                if mnemonic.startswith("vset"):
+                    group = "vset"
+                elif mnemonic.startswith("vl"):
+                    group = "load"
+                elif mnemonic.startswith("vs") and not mnemonic.startswith("vset"):
```
Comment (kathlenemagnus) [inline scripts/generate_vector_tests.py:199]: Did you find any loads or stores that were added to the load/store groups here? Meaning that the inst type in Mavis didn't included "load" or "store"? I am hoping lines 197-200 can be removed.

File: scripts/generate_vector_tests.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                elif any(x in mnemonic for x in ["seq", "slt", "sne", "sgt", "sge", "sle"]):
+                    group = "compare"
+                elif "slide" in mnemonic:
+                    group = "slide"
+                elif "gather" in mnemonic:
+                    group = "gather"
+                elif "permute" in mnemonic or ("mv" in mnemonic and not mnemonic.startswith("vset")):
+                    group = "permute"
+                else:
+                    group = "other"
```
Comment (kathlenemagnus) [inline scripts/generate_vector_tests.py:193]: The other test is very large... I would add more groups to reduce the size of this test (e.g. float, widening, mask gen).

**Comment by kathlenemagnus** [comment]
Hi @ADI-ROXX and @froglodyte, first I want to thank you both for tackling this problem! Unfortunately there is some duplicate work between your PRs (this one and #291), but I have a proposal. 

Both of your approaches to the problem have merit. @ADI-ROXX's script ensures that ALL vector instructions are represented in the tests, while @froglodyte's script tests different combinations of instructions with more randomness. Both are great strategies for testing Olympia's vector support.

What I'd like to propose moving forward is using this PR as a starting point and extending @ADI-ROXX's script to include some of the features from @froglodyte's PR:
- Use random operand numbers and immediate values
- Generate shorter tests with instructions randomly inserted. We can generate multiple tests for each group to ensure all instructions are still included.
- Use argparse to configure the script

There are also some features I'd like to add in a future PR:
- Provide VLEN as a parameter (will be used to determine max vector length)
- Use random vector lengths and vector configs for each instruction ('vset{i}vl{i} randomziation including whether vl == vlmax)
- Randomly inject dependencies between instructions

@ADI-ROXX and @froglodyte, please let me know what you think of my proposal.

**Comment by Froglodyte** [comment]
Hi @kathlenemagnus, those changes sound great. @ADI-ROXX will you work on them? I have exams coming up so I'll be a bit busy for the next two weeks.

**Comment by ADI-ROXX** [comment]
>@ADI-ROXX will you work on them? I have exams coming up so I'll be a bit busy for the next two weeks.

Yeah, sure. I will merge your changes in this pr. 
Sorry, I was on a vacation due to which I wasn't able to reply 😅

**Comment by klingaard** [comment]
@ADI-ROXX what's the status of this PR?  Can you get to it and see it to the end?

**Comment by ADI-ROXX** [comment]
> @ADI-ROXX what's the status of this PR? 

Trying to finish it by this week.

**Comment by ADI-ROXX** [comment]
Hi team, some updates from my side. I have merged Froglodyte's ideas in my PR. 

Presently, some of the tests are failing due to FetchQueue error. I am working on fixing them

**Comment by ADI-ROXX** [comment]
@kathlenemagnus @klingaard I have opened a new issue #302. Can you give it a look once. I think that the issue has to be resolved before merging of this PR.
